### PR TITLE
py: Add optional support for descriptors' __get__ and __set__ methods

### PR DESCRIPTION
**Pull request by stinos**
Disabled by default as it requires an extra mp_load_method_maybe for every attribute store

When enabled the cmdline/cmd_xxx tests fail. What is the policy for these - are they here to stay? If so is there a script to generate the exp file or does it have to be created manually after each change?

### CODE REVIEW

File: py/objtype.c [hunk 1]
```diff
@@ -535,6 +549,24 @@ bool mp_obj_instance_store_attr(mp_obj_t self_in, qstr attr, mp_obj_t value) {
         .is_type = false,
     };
     mp_obj_class_lookup(&lookup, self->base.type);
+#endif
+
+#if MICROPY_PY_BUILTINS_DESCRIPTOR
+    if (member[0] != MP_OBJ_NULL) {
+        // found a class attribute; if it has a __set__ method then call it with the
+        // class instance and valu as arguments
```
Comment (pfalcon) [inline py/objtype.c]: typo

**Comment by pfalcon** [comment]
Dunno, I wish there was an RFC on how to support descriptors first.

**Comment by dpgeorge** [comment]
> When enabled the cmdline/cmd_xxx tests fail. What is the policy for these - are they here to stay? If so is there a script to generate the exp file or does it have to be created manually after each change?

It fails because you added qstrs and so the bytecode is different (the qstr numbers for all those following the new ones are changed).  I'm working on a fix to this.

For testing purposes your new feature can be enabled on the unix port.

**Comment by stinos** [comment]
Yes I figured it's due to the qstrs. I updated the bytecode in the exp file, but tests continue to fail because somehow the .out file produced is different everytime (and not just in the locations marked with ######): the lines around

164 STORE_MAP
165 LOAD_FAST 14
166 LOAD_FAST 15

are always different.. For instance one run I'll get

164 BUILD_SET 1023970
168 LOAD_CONST_SMALL_INT -128
169 LOAD_CONST_SMALL_INT -128
170 LOAD_BUILD_CLASS
171 BINARY_OP 11 
172 LOAD_CONST_SMALL_INT 25
173 LOAD_CONST_SMALL_INT -3
174 DUP_TOP_TWO
175 LOAD_CONST_SMALL_INT -1
176 LOAD_CONST_SMALL_INT -128
177 LOAD_CONST_SMALL_INT -128
178 LOAD_CONST_SMALL_INT -126
179 LOAD_FAST 2

and the next it becomes

164 BUILD_SET 1023970
168 LOAD_CONST_SMALL_INT -128
169 LOAD_CONST_SMALL_INT -128
170 LOAD_BUILD_CLASS
171 BINARY_OP 27 __le__
172 UNARY_OP 3
173 LOAD_CONST_SMALL_INT -54
174 LOAD_CONST_SMALL_INT 40
175 LOAD_CONST_SMALL_INT -1
176 LOAD_CONST_SMALL_INT -128
177 LOAD_CONST_SMALL_INT -128
178 LOAD_CONST_SMALL_INT -126
179 LOAD_FAST 2

Does this mean there's effectively a bug?

@pfalcon what's special about descriptors is they have a place in the lookup chain for attribute access but that's about it. Micropython implements only part of that lookup chain anyway as far as I can tell, so in order to support descriptors it seems logical to put their lookup in the place which comes closest to what CPython does?

**Comment by pfalcon** [comment]
Re: showbc output different: I see the same in https://github.com/micropython/micropython/pull/1162

**Comment by pfalcon** [comment]
> what's special about descriptors is they have a place in the lookup chain for attribute access but that's about it.

Well, I'm not proficient with descriptors and never use them in real life. I had an idea that descriptors might help with following situation: in CPython, "os.path" is pre-imported when you do "import os". In uPy, "os" is package and "os.path" is a submodule inside it, not pre-imported (because it's waset of memory). I hoped that descriptors would allow me to make "path" some proxy object and do lazy-load on first access. Ahem, they can't do, on their own - it's specifically class instances attribute lookup which supports descriptor protocol. So, a usecase: people who never used descriptors found possible use for them, but it didn't work. Discuss: 1) are descriptors really useful then? 2) how to extend them to be useful for real-world internal tasks.

Then, reading https://docs.python.org/3/howto/descriptor.html , there're "data" and "non-data" descriptors, and lot of other blah-blah. The doc says "it creates ... an appreciation for the elegance of its design.", but I fail to see an elegance in bunch of those adhoc rules and it only fortifies opinion that descriptor is internal CPython implementation detail. Which leads to a usual question - why the patch is needed and why it is implemented the way it is?

Otherwise I surely agree that it's nice to be able to tick an item "descriptor support" in uPy's feature list, but let's make sure we all have agreement that this patch doesn't implement anything compatible with CPython, so it's just an experimental feature, and in particular, should not be enabled by default (like patch does, yeah - it's just a bit of testing/regression issue).

**Comment by stinos** [comment]
I have only one single usecase for it, but for me it's worth it, simplifies code and makes it safer to use. So I just submitted the patch since I have the code lying around anyway. Same goes for the function name PR btw. This is what I wanted:

```
class cached_property:
  def __init__( self, fget ):
    self.fget = fget
    self.func_name = fget.__name__

  def __get__( self, obj, cls ):
    if obj is None:
      return None
    value = self.fget( obj )
    setattr( obj, self.func_name, value )
    return value
```

The thing I'm trying to solve here is one of the top layers of a piece of software which should be fairly simple to use, also because it gets used by not-so-much-tech people (who have sometimes never programmed before). Practically there's a factory which creates objects based on some configuration file. However due to the nature of the objects created (for example AD/DA converter abstractions) most of them must be created only once and furthermore the order of creation matters (for example one can only create a Timer object for an AD/DA once the latter has been created). 

Now instead of trying to instruct all users to "only call this method once, then use the instance it creates and pass it to method x and y" , which they then won't do anyway and instead come nagging on my desk and wast my time, I instead use this now:

```
class DaqFactory:
  @cached_property
  def DaqDevice(self):
    #create AD/DA stuff
    return device

  @cached_property
  def Edgedetector(self):
    device= self.DaqDevice
    #create object using master and config and return it once
    return proc
```

so it doesn't matter if the user does 

```
detector = factory.Edgedetector
factory.DaqDevice.Start()
```

or

```
device = factory.DaqDevice
detector = factory.Edgedetector
device.Start()
```

or whatever he/she wants. And that is the goal exactly. There are probably other ways to do this, but I doubt they are as convenient and performant as this one.

**Comment by pfalcon** [comment]
@stinos: Thanks for these details, if your changes allow to do that (in CPython-compatible way), they're certainly good. (And no, by mere look at your code I don't understand how it works ;-) ).

**Comment by dpgeorge** [comment]
I agree that the way you have done it is the best way to fit it into the current code.  But it does introduce a large performance hit (well I guess it does, remains to be tested!).  Merged in 28fa84b4454ef253dfba5f1ceb023be2184a7271 because descriptors can be very powerful and we can improve on it in the future if needed.

**Pull request closed without merging**